cmd/multiFaAcceleration/multiFaAcceleration.go

```diff
@@ -0,0 +1,504 @@
+	} else if sub.vc < 0 {
+		sub.vc = 0
+		sub.va = nonNegativeApproximation(sub.Dab, sub.Dac, sub.vb, sub.vc, s)
+		sub.vb = nonNegativeApproximation(sub.Dab, sub.Dbc, sub.va, sub.vc, s)
+	}
+	return sub.va, sub.vb, sub.vc
+}
+
+//If we constrain branch lengths to be nonNegative, we apply this correction when the minimum Q is achieved at negative branch lengths for a subtree.
+func nonNegativeApproximation(d1 float64, d2 float64, v1 float64, v2 float64, s Settings) float64 {
+	if d1 == 0 {
+		if d2 == 0 {
+			return numbers.MaxFloat64(0, (s.ZeroDistanceWeightConstant*(d1-v1)+s.ZeroDistanceWeightConstant*(d2-v2))/(2*s.ZeroDistanceWeightConstant))
+		} else {
+			return numbers.MaxFloat64(0, s.ZeroDistanceWeightConstant*(d1-v1)+(1.0/math.Pow(d2, 2)*(d2-v2))) / (s.ZeroDistanceWeightConstant + (1.0 / math.Pow(d2, 2)))
+		}
+	} else if d2 == 0 {
+		return numbers.MaxFloat64(0, (1.0/(math.Pow(d1, 2))*(d1-v1)+s.ZeroDistanceWeightConstant*(d2-v2))/((1.0/math.Pow(d1, 2))+s.ZeroDistanceWeightConstant))
+	}
+	return numbers.MaxFloat64(0, (1.0/(math.Pow(d1, 2))*(d1-v1)+(1.0/math.Pow(d2, 2)*(d2-v2)))/((1.0/math.Pow(d1, 2))+(1.0/math.Pow(d2, 2))))
+}
+
+//Reduce the four species tree to the subtree containing species 0, 1, and the ancestor of 2/3.
+func pruneLeft(d Distances, b BranchLengths, sub *SubTree, s Settings) {
+	sub.Dab = d.D01
+	if d.D03 == 0 {
+		if d.D02 == 0 {
+			sub.Dac = (s.ZeroDistanceWeightConstant*(d.D02-b.B4) + s.ZeroDistanceWeightConstant*(d.D03-b.B5)) / (2 * s.ZeroDistanceWeightConstant)
+		} else {
+			sub.Dac = ((1.0/math.Pow(d.D02, 2))*(d.D02-b.B4) + (s.ZeroDistanceWeightConstant * (d.D03 - b.B5))) / (s.ZeroDistanceWeightConstant + (1.0 / math.Pow(d.D02, 2)))
+		}
+	} else if d.D02 == 0 {
+		sub.Dac = (s.ZeroDistanceWeightConstant*(d.D02-b.B4) + ((1.0 / math.Pow(d.D03, 2)) * (d.D03 - b.B5))) / ((1.0 / math.Pow(d.D03, 2)) + s.ZeroDistanceWeightConstant)
+	} else {
+		sub.Dac = ((1.0/math.Pow(d.D02, 2))*(d.D02-b.B4) + (1.0/math.Pow(d.D03, 2))*(d.D03-b.B5)) / ((1.0 / math.Pow(d.D03, 2)) + (1.0 / math.Pow(d.D02, 2)))
+	}
+
+	if d.D13 == 0 {
+		if d.D12 == 0 {
+			sub.Dbc = (s.ZeroDistanceWeightConstant*(d.D12-b.B4) + s.ZeroDistanceWeightConstant*(d.D13-b.B5)) / (2 * s.ZeroDistanceWeightConstant)
+		} else {
+			sub.Dbc = (s.ZeroDistanceWeightConstant*(d.D13-b.B5) + (1.0/math.Pow(d.D12, 2))*(d.D12-b.B4)) / ((1.0 / math.Pow(d.D12, 2)) + s.ZeroDistanceWeightConstant)
+		}
+	} else if d.D12 == 0 {
+		sub.Dbc = (s.ZeroDistanceWeightConstant*(d.D12-b.B4) + (1.0/math.Pow(d.D13, 2))*(d.D13-b.B5)) / ((1.0 / math.Pow(d.D13, 2)) + s.ZeroDistanceWeightConstant)
+	} else {
+		sub.Dbc = ((1.0/math.Pow(d.D12, 2))*(d.D12-b.B4) + (1.0/math.Pow(d.D13, 2))*(d.D13-b.B5)) / ((1.0 / math.Pow(d.D13, 2)) + (1.0 / math.Pow(d.D12, 2)))
+	}
+}
+
+//Reduce the four species tree to the subtree containing 2, 3, and the ancestor of 0/1.
+func pruneRight(d Distances, b BranchLengths, sub *SubTree, s Settings) {
+	sub.Dac = d.D23
+
+	if d.D02 == 0 {
```

Comment:
Similar to my above comment, I think all these long and separate equations makes it difficult to pick out bugs.  Can we make the weights variables, which get set with if-else statements and then we just have one equation?